### ISSUE TITLE
option_if_let_else suggests map_or instead of map_or_else even if else branch contains a function call

### ISSUE THREAD

**Issue by Arnavion** [rust-lang/rust-clippy#5821]
```rust
#[deny(clippy::option_if_let_else)]
pub fn foo(o: Option<i32>, f: impl FnOnce(i32), g: impl FnOnce()) {
    if let Some(i) = o {
        f(i)
    }
    else {
        g()
    }
}
```

The lint suggests `o.map_or(g(), |i| f(i))`, which is wrong because the `else` branch should be evaluated lazily because it's a function call. It ought to have suggested `o.map_or_else(|| g(), |i| f(i))`

### Meta

- `cargo clippy -V`: `clippy 0.0.212 (39d5a61 2020-07-17)`
- `rustc -Vv`:
  ```
  rustc 1.47.0-nightly (39d5a61f2 2020-07-17)
  binary: rustc
  commit-hash: 39d5a61f2e4e237123837f5162cc275c2fd7e625
  commit-date: 2020-07-17
  host: x86_64-unknown-linux-gnu
  release: 1.47.0-nightly
  LLVM version: 10.0
  ```

**Comment by Arnavion** [comment]
I now see that https://github.com/rust-lang/rust-clippy/pull/5301#issuecomment-597738782 implies this is intentional, and that after the suggestion, the `clippy::or_fun_call` lint would suggest changing it to `map_or_else`. However

```rust
#[deny(clippy::or_fun_call)]
pub fn foo(o: Option<i32>, f: impl FnOnce(i32), g: impl FnOnce()) {
    o.map_or(g(), |i| f(i))
}
```

does not raise any errors.

It does cause the `clippy::unit_arg` lint to fire, but that lint itself has a bad suggestion, for which I've filed #5823

So if we modify the code to remove the possibility of triggering `clippy::unit_arg`:

```rust
#[deny(clippy::or_fun_call)]
pub fn foo(o: Option<i32>, f: impl FnOnce(i32) -> i32, g: impl FnOnce() -> i32) -> i32 {
    o.map_or(g(), |i| f(i))
}
```

this code still does not raise any errors.

Perhaps this issue can be repurposed to figure out why `clippy::or_fun_call` did not fire on the code after applying `clippy::if_let_else_`'s suggestion.

**Comment by montrivo** [comment]
I think `option_if_let_else` should recommend `map_or_else` by default to preserve the code semantic. If the else block consists in a single identifier (or another pattern without side-effect), then `map_or` is applicable.

**Comment by montrivo** [comment]
```
#[deny(clippy::or_fun_call)]
pub fn foo(o: Option<i32>, f: impl FnOnce(i32) -> i32, g: impl FnOnce() -> i32) -> i32 {
    o.map_or(g(), |i| f(i))
}
```
doesn't trigger the lint because it only looks for function calls of at least one argument. See line 1752 `let or_has_args = !or_args.is_empty();`
The reasoning was maybe that functions taking no argument have probably side-effects.

**Comment by xFrednet** [comment]
This issue seems to be fixed. Clippy at least suggests the usage of `map_or_else` for the example code. [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=54b3a96ad565c52c43756804d956e5c3)

**Comment by Arnavion** [comment]
The change you're seeing is from https://github.com/rust-lang/rust-clippy/pull/5937 , which considered this issue to only be "partially fixed". I'm not sure what the unfixed part is, however.

**Comment by J-ZhengLi** [comment]
as of today, this does not lint at all, not sure when that happened or whether it was on purpose or not

**Comment by Hasnabruzzn** [comment]
> as of today, this does not lint at all, not sure when that happened or whether it was on purpose or not

As f(i32) and g() return unit, #11893 / #11896 apply

If you make the closures return a value the lint **is** triggered:
```
#[warn(clippy::option_if_let_else)]
fn closures_return_value(o: Option<i32>, f: impl FnOnce(i32) -> i32, g: impl FnOnce() -> i32) -> i32 {
    if let Some(i) = o {
        f(i)
    } else {
        g()
    }
}
```

**Comment by Arnavion** [comment]
And for the version that returns a value it does correctly suggest `map_or_else(g)` instead of `map_or(g())`, so there doesn't seem to be anything left here to fix.
